solution/ui/prototype/src/utilities/api.js

```diff
@@ -0,0 +1,371 @@
+                            );
+                    }
+                    throw parseError({
+                        message: text,
+                        status: httpStatus,
+                    });
+                } else {
+                    throw parseError(
+                        new Error("The server did not return a json response.")
+                    );
+                }
+            }
+
+            return json;
+        })
+        .then((json) => {
+            if (_isBoolean(isOk) && !isOk) {
+                throw parseError({ ...json, status: httpStatus });
+            } else {
+                return json;
+            }
+        });
+}
+
+// ---------- helper functions ---------------
+
+function httpApiMock(verb, urlPath, { data, params, response } = {}) {
+    console.log(`Performing an HTTP ${verb} to ${config.apiPath}/${urlPath}`);
+    data && console.log("DATA: ", data);
+    params && console.log("PARAMS: ", params);
+    response && console.log("RESPONSE: ", response);
+    return response;
+}
+
+function httpApiGet(urlPath, { params } = {}) {
+    return fetchJson(`${config.apiPath}/${urlPath}`, {
+        method: "GET",
+        headers: authHeader(token),
+        params,
+    });
+}
+
+function httpApiPost(urlPath, { data = {}, params } = {}) {
+    console.log(data);
+    return fetchJson(`${config.apiPath}/${urlPath}`, {
+        method: "POST",
+        headers: authHeader(token),
+        params,
+        body: JSON.stringify(data),
+    });
+}
+
+// eslint-disable-next-line no-unused-vars
+function httpApiPut(urlPath, { data, params } = {}) {
+    return fetchJson(`${config.apiPath}/${urlPath}`, {
+        method: "PUT",
+        headers: authHeader(token),
+        params,
+        body: JSON.stringify(data),
+    });
+}
+
+// eslint-disable-next-line no-unused-vars
+function httpApiDelete(urlPath, { data, params } = {}) {
+    return fetchJson(`${config.apiPath}/${urlPath}`, {
+        method: "DELETE",
+        headers: authHeader(token),
+        params,
+        body: JSON.stringify(data),
+    });
+}
+
+// ---------- api calls ---------------
+
+const login = (username, password) => {
+    return httpApiPost("Authentication/login", {
+        data: { username, password },
+    });
+};
+
+const statusCheck = () => {
+    return httpApiGet("Authentication/Test");
+};
+
+const getLastUpdatedDate = async (title = "42") => {
+    const reducer = (accumulator, currentValue) => {
+        return currentValue.date > accumulator.date
+            ? currentValue
+            : accumulator;
+    };
+
+    const result = await httpApiGet(`title/${title}/existing`);
+
+    return niceDate(_get(result.reduce(reducer), "date"));
+};
+
+const getPartNames = async (title = "42") => {
+    const result = await httpApiGet(`title/${title}/existing`);
+
+    return _sortedUniq(result.flatMap((part) => part.partName).sort());
+};
+
+const getHomepageStructure = async () => {
```

Comment:
Can this reducer be shelled out into a utils class?  It does some post processing of the result to make it into a more usable format right?  Should we do more of this in Django or in the parser anyways?

Reply:
we're totally outside of Django here so this would have to be something the parser does of the front end does.  

Caleb and I talked about using the parser to create a bunch of new data structures every time it runs, structures that would be specific to certain endpoints or parts of the prototype.  So yes, the parser would be a better place to do this transformation.  Caleb and I talked about some future stories that could/would be written to do exactly that.

As far as where to put this - I think this could be put into a utils class for sure, broken up into parts that can be composed together.  Although I don't think now is the right time to do that because we don't know what parts of this reducer would be helpful elsewhere.

However, I've started work on EREGCSC-1182 - building out the JS part view - and I'm seeing likely areas of overlap.  So this would be work better suited for that story.